File: php/scripts/update_coverage_badge.php
<?php

declare(strict_types=1);

if ($argc !== 3) {
    fwrite(STDERR, "usage: php scripts/update_coverage_badge.php <clover.xml> <readme.md>\n");
    exit(2);
}

[$_, $coverageFile, $readmePath] = $argv;
$xml = simplexml_load_file($coverageFile);
if ($xml === false) {
    fwrite(STDERR, "failed to read coverage file: {$coverageFile}\n");
    exit(2);
}

$metrics = $xml->project->metrics;
$covered = (int)$metrics['coveredstatements'];
$statements = (int)$metrics['statements'];
if ($statements === 0) {
    fwrite(STDERR, "coverage file has no statements\n");
    exit(2);
}

$percent = (int)floor(($covered / $statements) * 100);
$color = match (true) {
    $percent >= 90 => 'brightgreen',
    $percent >= 80 => 'green',
    $percent >= 70 => 'yellowgreen',
    $percent >= 60 => 'yellow',
    $percent >= 50 => 'orange',
    default => 'red',
};

$readme = file_get_contents($readmePath);
if ($readme === false) {
    fwrite(STDERR, "failed to read README: {$readmePath}\n");
    exit(2);
}

$pattern = '~https://img\.shields\.io/badge/coverage-[0-9.]+%25-[a-z]+~';
$replacement = "https://img.shields.io/badge/coverage-{$percent}%25-{$color}";
$updated = preg_replace($pattern, $replacement, $readme, -1, $count);
if ($updated === null || !is_string($updated)) {
    fwrite(STDERR, "preg_replace failed against README badge\n");
    exit(2);
}
if ($count === 0) {
    fwrite(STDERR, "no coverage badge found in {$readmePath}\n");
    exit(1);
}

if ($updated === $readme) {
    printf("Coverage badge already at %d%% (%s). No change.\n", $percent, $color);
    exit(0);
}

if (file_put_contents($readmePath, $updated) === false) {
    fwrite(STDERR, "failed to write README: {$readmePath}\n");
    exit(2);
}

printf("Updated coverage badge: %d%% (%s)\n", $percent, $color);

File: ruby/scripts/update_coverage_badge.rb
# frozen_string_literal: true

require "json"

unless ARGV.length == 2
  warn "usage: ruby scripts/update_coverage_badge.rb <simplecov-resultset.json> <readme.md>"
  exit 2
end

resultset_path, readme_path = ARGV
resultset = JSON.parse(File.read(resultset_path))
coverage = resultset.values.fetch(0).fetch("coverage")

line_total = 0
line_covered = 0
branch_total = 0
branch_covered = 0

coverage.each_value do |metrics|
  metrics.fetch("lines", []).each do |hits|
    next if hits.nil?

    line_total += 1
    line_covered += 1 if hits.positive?
  end

  metrics.fetch("branches", {}).each_value do |branches|
    branches.each_value do |hits|
      branch_total += 1
      branch_covered += 1 if hits.positive?
    end
  end
end

def percentage(covered, total)
  raise "coverage result has no measurable items" if total.zero?

  ((covered.to_f / total) * 100).floor
end

def badge_color(percent)
  case percent
  when 90.. then "brightgreen"
  when 80...90 then "green"
  when 70...80 then "yellowgreen"
  when 60...70 then "yellow"
  when 50...60 then "orange"
  else "red"
  end
end

line_percent = percentage(line_covered, line_total)
branch_percent = percentage(branch_covered, branch_total)
readme = File.read(readme_path)
updated = readme
updated = updated.gsub(
  %r{https://img\.shields\.io/badge/coverage-[0-9.]+%25-[a-z]+},
  "https://img.shields.io/badge/coverage-#{line_percent}%25-#{badge_color(line_percent)}"
)
updated = updated.gsub(
  %r{https://img\.shields\.io/badge/branch%20coverage-[0-9.]+%25-[a-z]+},
  "https://img.shields.io/badge/branch%20coverage-#{branch_percent}%25-#{badge_color(branch_percent)}"
)

if updated == readme
  puts "Coverage badges already at #{line_percent}% line / #{branch_percent}% branch."
else
  File.write(readme_path, updated)
  puts "Updated coverage badges: #{line_percent}% line / #{branch_percent}% branch."
end
